website/api/controllers/create-or-update-one-newsletter-subscription.js

```diff
@@ -0,0 +1,59 @@
+module.exports = {
+
+
+  friendlyName: 'Create or update one newsletter subscription',
+
+
+  description: 'Creates or updates a NewsletterSubscription record for the provided email address',
+
+
+  inputs: {
+    emailAddress: {
+      type: 'string',
+      required: true,
+    },
+
+    subscribeTo: {
+      type: 'string',
+      required: true,
+      description: 'The type of content that this user is changing their subscription for',
+      isIn: ['releases']
+    }
+  },
+
+
+  exits: {
+
+    success: {
+      description: 'A user has successfully changed their subscription to the Fleet newsletter'
+    },
+
+  },
+
+
+  fn: async function ({emailAddress, subscribeTo}) {
+
+    // Look for an existing NewsletterSubscription that uses the provided email address
+    let doesSubscriptionForProvidedEmailExist = await NewsletterSubscription.findOne({emailAddress: emailAddress});
+
+    // If one does not exist, we'll create a new one.
+    if(!doesSubscriptionForProvidedEmailExist) {
+      await NewsletterSubscription.create({emailAddress: emailAddress});
+    }
```

Comment:
Forgot to mention this last time.  It's minor and doesn't come up often, but a best practice so we might as well do it.

Instead of finding, then creating, can you change this to first attempt a create, and if we get a uniqueness constraint violation, _only then_ look up the existing subscription?

[The reason why 🥛 »](https://web.stanford.edu/~ouster/cgi-bin/cs140-winter16/lecture.php?topic=concurrency#:~:text=The%20%22Too%20Much%20Milk%22%20Problem&text=More%20definitions%3A,executing%20at%20a%20given%20time.)

Reply:
That makes sense, thank you!